code/controllers/subsystem/input.dm

```diff
@@ -36,4 +36,34 @@ SUBSYSTEM_DEF(input)
 
 /datum/controller/subsystem/input/fire()
 	for(var/mob/user as anything in GLOB.keyloop_list)
-		user.focus?.keyLoop(user.client)
+		if(user.focus)
+			var/movement_dir = NONE
+			for(var/_key in user.client?.keys_held)
+				movement_dir = movement_dir | user.movement_keys[_key]
+			if(user?.next_move_dir_add)
+				movement_dir |= user.client.next_move_dir_add
+			if(user?.next_move_dir_sub)
+				movement_dir &= ~user.client.next_move_dir_sub
```

Comment:
I'm reasonably sure all these var accesses outweigh any benefit of unrolling the loop. A proc with no args isn't actually that expensive to call. This also means that any runtimes in keyLoop break ssinput, which is worrying.

Shouldn't need to run this on live to profile it, have you checked yourself?

Reply:
The proc had an arg, it fed in the client.

Reply:
oh that's true. I wonder if it's still faster in just the base case of input firing. Would be good to see some clean profiles